File: python/1-50/23. Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not lists:
            return None
        end = len(lists) - 1
        while end > 0:
            start = 0
            while start < end:
                lists[start] = self.merge(lists[start], lists[end])
                start += 1
                end -= 1
        return lists[0]
    def merge(self, a, b):
        if not a:
            return b
        if not b:
            return a
        head = ListNode(0)
        cur = head
        while a and b:
            if a.val < b.val:
                cur.next = a
                a = a.next
            else:
                cur.next = b
                b = b.next
            cur = cur.next
        cur.next = a or b
        return head.next

File: cpp/1-50/23. Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty())
            return NULL;
        return helper(lists, 0, lists.size() - 1);
    }
    ListNode* helper(vector<ListNode*>& lists, int low, int high)
    {
        if (low > high)
            return NULL;
        if (low == high)
            return lists[low];
        int mid = (low + high) / 2;
        ListNode *left = helper(lists, low, mid);
        ListNode *right = helper(lists, mid + 1, high);
        return merge(left, right);
    }
    ListNode* merge(ListNode *l1, ListNode *l2)
    {
        ListNode *dummyhead = new ListNode(0);
        ListNode *ptr = dummyhead;
        while (l1 && l2)
        {
            ptr->next = (l1->val < l2->val) ? l1 : l2;
            ptr = ptr->next;
            if (l1->val < l2->val)
                l1 = l1->next;
            else
                l2 = l2->next;
        }
        ptr->next = l1 ? l1 : l2;
        return dummyhead->next;
    }
};